pkg/parser/parser.go

```diff
@@ -1,121 +1,77 @@
-	GraphqlEndpoint string
+	// RPCEndpoint specifies an alternative RPC endpoint to use. If empty, DefaultRPCEndpoint is used.
+	RPCEndpoint string
 }
 
 const (
-	DefaultGraphqlEndpoint = "https://api.cirrus-ci.com/graphql"
-	defaultTimeout         = time.Second * 5
+	DefaultRPCEndpoint = "us-central1.gcp.api.cirrus-ci.com:8239"
+	defaultTimeout     = time.Second * 5
 )
 
 type Result struct {
 	Errors []string
 }
 
-const gqlQuery = `mutation Validate($clientMutationId: String, $config: String) {
-  validate(input: {clientMutationId: $clientMutationId, config: $config}) {
-    errors
-  }
-}
-`
-
-type gqlValidate struct {
-	Errors []string
-}
-
-type gqlData struct {
-	Validate gqlValidate
-}
-
-type gqlError struct {
-	Message string
-}
-
-type gqlResponse struct {
-	Data   *gqlData
-	Errors []gqlError
-}
-
-func (p *Parser) graphqlEndpoint() string {
-	if p.GraphqlEndpoint == "" {
-		return DefaultGraphqlEndpoint
+func (p *Parser) rpcEndpoint() string {
+	if p.RPCEndpoint == "" {
+		return DefaultRPCEndpoint
 	}
 
-	return p.GraphqlEndpoint
+	return p.RPCEndpoint
 }
 
 func (p *Parser) Parse(config string) (*Result, error) {
-	// Prepare GraphQL query
-	query := struct {
-		Query     string            `json:"query"`
-		Variables map[string]string `json:"variables"`
-	}{
-		Query: gqlQuery,
-		Variables: map[string]string{
-			"clientMutationId": uuid.New().String(),
-			"config":           config,
-		},
-	}
-	buf, err := json.Marshal(&query)
-	if err != nil {
-		return nil, fmt.Errorf("%w, %v", ErrInternal, err)
-	}
-	requestBody := bytes.NewBuffer(buf)
+	// Create a context to enforce the defaultTimeout
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
 
-	// Send config for validation via Cirrus CI API
-	client := http.Client{Timeout: defaultTimeout}
-	resp, err := client.Post(p.graphqlEndpoint(), "text/json", requestBody)
+	// Setup Cirrus CI RPC connection
+	conn, err := grpc.DialContext(ctx, p.rpcEndpoint(), grpc.WithInsecure())
 	if err != nil {
-		return nil, fmt.Errorf("%w, %v", ErrTransport, err)
+		return nil, fmt.Errorf("%w: %v", ErrRPC, err)
 	}
-	defer resp.Body.Close()
+	defer conn.Close()
 
-	// GraphQL's convention is to always return HTTP 200,
-	// in all other cases something has really gone wrong.
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("%w: got HTTP %d (%s)", ErrTransport, resp.StatusCode, resp.Status)
-	}
+	// Send config for parsing by the Cirrus CI RPC and retrieve results
+	client := api.NewCirrusCIServiceClient(conn)
 
-	responseBody, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, fmt.Errorf("%w: %v", ErrTransport, err)
+	request := api.ParseConfigRequest{
+		Config: config,
 	}
 
-	// Parse and validate JSON result
-	var parsed gqlResponse
-	err = json.Unmarshal(responseBody, &parsed)
+	_, err = client.ParseConfig(ctx, &request)
```

Comment:
The response now contains parsed tasks. Maybe we can follow up this PR with one to have a flag to print these tasks out? Or maybe globally support `--verbose` flag to have verbose logging of everything. 🤔

Reply:
That would be a nice and logical addition indeed, however it seems that these details would be more applicable to the command that actually executes the tasks (e.g. `cirrus run`). Perhaps we should implement it first, and then the answer will be more evident?